### PR TITLE
Remove the temporary mechanism for "external" nullable annotations.

### LINKED ISSUE #34409: Remove hard-coded nullable annotations in compiler before C# 8.0 RTM

### PR DESCRIPTION
**Pull request by gafter**
Relates to #29821
Fixes #34409

### CODE REVIEW

**Comment by gafter** [comment]
Not sure if this is the right time to do this.  See for example https://github.com/dotnet/roslyn/issues/34901

**Comment by jcouv** [comment]
Fixes https://github.com/dotnet/roslyn/issues/34409 ?

File: src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs [hunk 1]
```diff
@@ -1,181 +0,0 @@
...
-                { "System.String System.String.Concat(System.String, System.String)", Array(Nullable(false), Nullable(true), Nullable(true)) },
-            }.ToImmutableDictionary();
-
-        private static readonly ImmutableDictionary<string, ImmutableArray<FlowAnalysisAnnotations>> Attributes =
-            new Dictionary<string, ImmutableArray<FlowAnalysisAnnotations>>
-            {
-                { "System.Boolean System.String.IsNullOrEmpty(System.String)", Array(default, NotNullWhenFalse) },
-                { "System.Boolean System.String.IsNullOrWhiteSpace(System.String)", Array(default, NotNullWhenFalse) },
-                { "System.Boolean System.String.Contains(System.String)", Array(default, EnsuresNotNull) },
-                { "System.Void System.Diagnostics.Debug.Assert(System.Boolean)", Array(default, AssertsTrue) },
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs]: >System.Void System.Diagnostics.Debug.Assert(System.Boolean)", Array(default, AssertsTrue) [](start = 19, length = 89)

Should the compiler treat `Debug.Assert(bool condition, ...)` as `Debug.Assert([AssertsTrue] bool condition, ...)`, even when unannotated?

**Comment by gafter** [comment]
I'm not going to merge this PR for now, until and unless we have a plan about what we will or will not do in this area.  See also #34901

**Comment by jcouv** [comment]
@gafter This PR could be refreshed and move ahead at this point. On the other hand, it is probably not crucial to do before RTM.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -24349,58 +24245,7 @@ void Main(string? c, bool b)
         }
 
         [Fact]
-        public void DoesNotReturnIfFalse_Debug_Assert2_NotNull()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: @cston was suggestion that we keep a hard-coded understanding of `Debug.Assert`.
I'm wonder if that's useful. Maybe we could keep those tests, but replace `Assert` with a custom `MyAssert` that is annotated with `DoesNotReturnIf(false)`?

**Comment by gafter** [comment]
@jcouv This has merged from master and changed to account for that merge since you reviewed it.  Are you OK with it still?

File: src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs [hunk 3]
```diff
@@ -71,7 +71,7 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
             while (stack.Count > 0);
 
             Debug.Assert(currentBinary != null);
-            return currentBinary;
+            return currentBinary!;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs]: Is this part of the change?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs [hunk 4]
```diff
@@ -687,8 +687,7 @@ internal override bool MustCallMethodsDirectly
                 var ordinal = i - 1;
                 bool isBad;
 
-                // https://github.com/dotnet/roslyn/issues/29821: handle extra annotations
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs]: I assume you checked there are no references left to this issue ID

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 5]
```diff
@@ -1739,7 +1739,7 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
                 return ImmutableArray.Create<DocumentId>();
             }
 
-            return _filePathToDocumentIdsMap.TryGetValue(filePath, out var documentIds)
+            return _filePathToDocumentIdsMap.TryGetValue(filePath!, out var documentIds)
```
Comment (jcouv) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs]: is this part of the change?

**Comment by jcouv** [comment]
The skipped test could simply be fixed by replacing `Debug.Assert` with a custom `MyAssert` with right annotation.

File: src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs]: No. That is the whole point of this change.

---
In reply to: [274941794](https://github.com/dotnet/roslyn/pull/34916#discussion_r274941794) [](ancestors = 274941794)

File: src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs]: Yes.  Apparently the version of `Debug.Assert` used here is not annotated.

---
In reply to: [303279357](https://github.com/dotnet/roslyn/pull/34916#discussion_r303279357) [](ancestors = 303279357)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs (hunk 4 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs]: Yes

---
In reply to: [303279434](https://github.com/dotnet/roslyn/pull/34916#discussion_r303279434) [](ancestors = 303279434)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 5 above)
Comment (gafter) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs]: Yes.  Apparently the version of `string.IsNullOrEmpty` used here is not annotated.

---
In reply to: [303279493](https://github.com/dotnet/roslyn/pull/34916#discussion_r303279493) [](ancestors = 303279493)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Do you believe we have insufficient test coverage for `DoesNotReturnIf(false)`?

---
In reply to: [303224415](https://github.com/dotnet/roslyn/pull/34916#discussion_r303224415) [](ancestors = 303224415)

**Comment by gafter** [comment]
Do you believe we have insufficent test coverage for that annotation?

---
In reply to: [511264736](https://github.com/dotnet/roslyn/pull/34916#issuecomment-511264736) [](ancestors = 511264736)

**Pull request merged**